### PR TITLE
ENH: Adding additional keywords to read_html for #13461

### LINKED ISSUE #13461: Feature Request: Expose more parsing options in html_read
I am trying to parse some HTML and would love to have `converters` (https://github.com/pydata/pandas/blob/master/pandas/io/parsers.py#L1582-L1586) exposed to the `read_html` function.  

Is there a reason why more parsing/all parsing options aren't exposed to `TextParser` at https://github.com/pydata/pandas/blob/master/pandas/io/html.py#L631-L634 ?  

If not, would it be of interest to have a PR that adds this functionality?

### PR DESCRIPTION
**Pull request by gte620v**
- [x] closes #13461
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@gte620v The suggestion in #10534 was to add `dtype` argument (although the converters keyword will be able to achieve a similar goal). Can you have a look if it is easy to pass through that keyword as well?

In the same line, I think it would be useful to at once evaluate the other keywords to see which ones could be useful for html parsing.

File: pandas/io/tests/test_html.py [hunk 1]
```diff
@@ -694,6 +694,60 @@ def test_bool_header_arg(self):
...
+                            <td> 0.233</td>
+                            </tr>
+                            <tr>
+                            <th>R_l0_g1</th>
+                            <td> 0.244</td>
+                            <td> 0.285</td>
+                            </tr>
+                        </tbody>
+                    </table>"""
+        raw_data = np.array([[u'R_l0_g0', '0.763', 0.233],
```
Comment (sinhrks) [inline pandas/io/tests/test_html.py]: pls compare with expected `DataFrame`.

**Comment by gte620v** [comment]
> Can you have a look if it is easy to pass through that keyword as well?

My initial impression is that dtype can not be easily dropped in since it is not supported by the python engine parser: https://github.com/pydata/pandas/blob/master/pandas/io/parsers.py#L463-L470

For reasons I haven't poked around enough to understand, the `read_html` function uses the `FileParser` function which uses the python engine: https://github.com/pydata/pandas/blob/master/pandas/io/parsers.py#L1600-L1601


>In the same line, I think it would be useful to at once evaluate the other keywords to see which ones could be useful for html parsing. 

Sure, I'll take a look.  It seems a bit cumbersome having to copy over args in 3 or 4 places to add them to `read_html`.  I poke through the code to see if there is a better way to add all applicable args.  Even without finding a better way to do things, it should be straightforward to use all of the `FileParser` args.  

Please let me know if you have any suggestions.

**Comment by gte620v** [comment]
It would be easiest to just pass around kwargs from `TextParser` in https://github.com/pydata/pandas/blob/master/pandas/io/parsers.py#L1546-L1601

Is there a better way to do that other than explicitly calling out the arguments in each of these places?
- https://github.com/pydata/pandas/blob/0c6226cbbc319ec22cf4c957bdcc055eaa7aea99/pandas/io/html.py#L719-L721
- https://github.com/pydata/pandas/blob/0c6226cbbc319ec22cf4c957bdcc055eaa7aea99/pandas/io/html.py#L743-L750
- https://github.com/pydata/pandas/blob/0c6226cbbc319ec22cf4c957bdcc055eaa7aea99/pandas/io/html.py#L757-L760

**Comment by gte620v** [comment]
@jorisvandenbossche I added support for a few more arguments: `keep_default_na`, `squeeze`, and `date_parser`.  

I also consolidated the argument definitions from being called in several places to only being enumerated in the `read_html` function.  This should make it easier to add more arguments to that function in the future.

I could not easily add dtype support since read_html uses the python engine.

Let me know if you want me to make more changes.

File: doc/source/whatsnew/v0.19.0.txt [hunk 2]
```diff
@@ -207,6 +207,12 @@ Other enhancements
 - The ``pd.read_csv()`` with ``engine='python'`` has gained support for the ``na_filter`` option (:issue:`13321`)
 - The ``pd.read_csv()`` with ``engine='python'`` has gained support for the ``memory_map`` option (:issue:`13381`)
 
+- The ``pd.read_html()`` has gained support for the ``na_values`` option (:issue:`13461`)
+- The ``pd.read_html()`` has gained support for the ``converters`` option (:issue:`13461`)
+- The ``pd.read_html()`` has gained support for the ``keep_default_na`` option (:issue:`13461`)
+- The ``pd.read_html()`` has gained support for the ``squeeze`` option (:issue:`13461`)
+- The ``pd.read_html()`` has gained support for the ``date_parser`` option (:issue:`13461`)
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v0.19.0.txt]: Can you combine this in one entry?

File: pandas/io/html.py [hunk 3]
```diff
@@ -716,9 +713,9 @@ def _validate_flavor(flavor):
     return flavor
 
 
-def _parse(flavor, io, match, header, index_col, skiprows,
-           parse_dates, tupleize_cols, thousands, attrs, encoding,
-           decimal):
+def _parse(flavor, io, match,
+           attrs, encoding,
+           **kwargs):
```
Comment (jorisvandenbossche) [inline pandas/io/html.py]: Can you put those on one line? (not needed for PEP8 to put on multiple lines here)

**Comment by jorisvandenbossche** [comment]
@gte620v The way of passing through looks good! Thanks

Regarding the extra keywords, maybe we should evaluate each keyword on its usefulness to add (sorry if I was a bit fast in pushing you to add more keywords):

- `keep_default_nan` -> certainly good to add (needed if we also add `na_values`, which was the original issue)
- `squeeze` / `date_parser` -> those are maybe less essential (the same effect can be easily obtained after parsing), so is maybe not worth the additional clutter in the long list of possible arguments

OK regarding not adding ``dtype`` for now (it's indeed not supported with the python engine), but can you then remove the note that it closed that issue?

**Comment by gte620v** [comment]
Sure, will do!

Let me know definitively about squeeze and date_parser.  I can leave or remove.

**Comment by jorisvandenbossche** [comment]
@jreback Opinion on adding `squeeze` and `date_parser` keywords to `to_html`? 
("it's easy to support" (just passing through to TextParser) vs "not essential keywords so only cluttering long list of possible keywords" (in contrast to the NaN handling keywords, the effect of `squeeze` and `date_parser` are easily obtained manually after parsing))

**Comment by jreback** [comment]
``squeeze`` should not be added; its purpose is to return a Series; while this method returns a list of dataframes (*always*), I would raise if its not-None.

``date_parser`` is kind of useless. you already have ``parse_dates`` to auto parse most dates, otherwise you use ``.to_datetime`` (usually with a format). To be honest I would blow that away from the csv parsers, but that's another thing. So raise here if its passed as well (a ``ValueError``)

**Comment by gte620v** [comment]
@jreback Thanks.  I removed squeeze and date_parser.  I think it is good to go.  Let me know if I misunderstood.

File: doc/source/whatsnew/v0.19.0.txt [hunk 4]
```diff
@@ -207,6 +207,8 @@ Other enhancements
 - The ``pd.read_csv()`` with ``engine='python'`` has gained support for the ``na_filter`` option (:issue:`13321`)
 - The ``pd.read_csv()`` with ``engine='python'`` has gained support for the ``memory_map`` option (:issue:`13381`)
 
+- The ``pd.read_html()`` has gained support for the ``na_values``, ``converters``, ``keep_default_na``  options (:issue:`13461`)
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: do we need an update in the docs themsleves? e.g. an example

**Comment by gte620v** [comment]
Sure, I can do that.

File: doc/source/whatsnew/v0.19.0.txt (hunk 4 above)
Comment (gte620v) [inline doc/source/whatsnew/v0.19.0.txt]: Done: added to io.rst

**Comment by gte620v** [comment]
I added examples to the docs.  Let me know if you want a modification or if I need to add something someplace else.

File: doc/source/io.rst [hunk 5]
```diff
@@ -1959,6 +1959,29 @@ Specify an HTML attribute
    dfs2 = read_html(url, attrs={'class': 'sortable'})
    print(np.array_equal(dfs1[0], dfs2[0]))  # Should be True
 
+Specify values that should be converted to NaN
+
```
Comment (jreback) [inline doc/source/io.rst]: can you add a version added tag for these

**Comment by jreback** [comment]
minor comment. ping when green.

**Comment by jorisvandenbossche** [comment]
@gte620v Thanks a lot!

**Comment by gte620v** [comment]
My pleasure!

**Pull request merged**